### ISSUE TITLE
Same padding in causal convolution provides wrong results

### ISSUE THREAD

**Issue by Atze00** [tensorflow/models#10062]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x ] I am using the latest TensorFlow Model Garden release and TensorFlow 2.
- [x ] I am reporting the issue to the correct repository. (Model Garden official or research directory)
- [x ] I checked to make sure that this issue has not been filed already.

## 1. The entire URL of the file you are using
https://github.com/tensorflow/models/blob/927e31aa1de2d23fd62b7b2644b67b29d658b944/official/vision/beta/modeling/layers/nn_layers.py

## 2. Describe the bug
1. Same padding in causal convolutions can leave some of the input unattended, look at the section 3 for more details.
2. <s> Same padding can be achieved padding either left or right, bottom or up. it appears that the implementation is different in the case of causal convolution with respect to standard convolution. The standard implementation starts from right, bottom, the causal convolution starts from left, top.</s> In some scenarios does both padding left and right, bottom and top, when just padding bottom right is the right choice.

## 3. Steps to reproduce
Same padiding leaves some of the input unattended:
```python
from official.vision.beta.modeling.layers.nn_layers import Conv3D 
kernel_size=(3,3,3)
stride=(1,3,3)
tf_conv = Conv3D( 1, kernel_size,kernel_initializer='ones', strides=stride, padding='causal',
                                                        use_bias=False)

a = tf.ones([1, 1, 3, 3, 1])
tf_conv(a)
```

Same padiding behaves differently<s>, padding left and top</s>, padding left right, bottom top:
```python
kernel_size=(3,3,3)
stride=(1,2,2)
tf_conv = Conv3D( 1, kernel_size,kernel_initializer='ones', strides=stride, padding='causal',
                                                        use_bias=False)

a = tf.ones([1, 1, 4, 4, 1])
tf_conv(a)
```

## 4. Expected behavior
First output should have been :
```<tf.Tensor: shape=(1, 1, 1, 1, 1), dtype=float32, numpy=array([[[[[9.]]]]], dtype=float32)>```
instead the output is: (this leaves the bottom right part of the input unattended)
```<tf.Tensor: shape=(1, 1, 1, 1, 1), dtype=float32, numpy=array([[[[[4.]]]]], dtype=float32)> ```

Second output should have been :
```<tf.Tensor: shape=(1, 1, 2, 2, 1), dtype=float32, numpy=
 array([[[[[9.],
           [6.]],
 
          [[6.],
           [4.]]]]], dtype=float32)> 
```
expected padding: 
```
[[0, 0], [2, 0], [0, 1], [0, 1], [0, 0]]
```
actual padding:
```
[[0, 0], [2, 0], [1, 1], [1, 1], [0, 0]]
```
instead the output is: 
```<tf.Tensor: shape=(1, 1, 2, 2, 1), dtype=float32, numpy=
 array([[[[[4.],
           [6.]],
 
          [[6.],
           [9.]]]]], dtype=float32)> 
```

## 5. Additional context

Code responsible for this behavior:  https://github.com/tensorflow/models/blob/dcdd2e4015b1c2cccc8e1345f9dc3f42df65d667/official/vision/beta/modeling/layers/nn_layers.py#L708
To the best of my knowledge it is not possible to create same padding without taking in consideration the input dimension.

## 6. System information
Note: the system is a colab notebook. Information like Bazel version is unknown to me.
-OS platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.5.0
- Python version:  3.7.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source): c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
- CUDA/cuDNN version: --
- GPU model and memory: --

**Comment by Atze00** [comment]
I can work on this issue if you confirm this.

**Comment by saikumarchalla** [comment]
@Atze00  Please submit a PR  once you aware of the changes required to fix the issue.Thanks!

**Comment by Atze00** [comment]
I'm working on it. I have a fix, I need to work on the tests. 
The models trained with this operator won't work as expected after the fix, are they going to be retrained?

**Comment by Hyperparticle** [comment]
Thanks for looking into this! From what I gather, this only happens when the spatial stride is greater than one?

I will check to see if this impacts the accuracy of our pretrained models. I'm guessing it might only have minor impact considering that the accuracy was similar to using `tf.keras.layers.Conv3D`.

**Comment by Atze00** [comment]
Looking at the code in the PR, you can obtain the current code substituting self.stride[i] with 1, so I guess for strides==1 the code works fine. There shouldn't be a difference in accuracy except in some cases (not sure which though) where you attend to only part of the input. Usually that happens with high strides value so they may not be effecting the models.
In case of attending to all the input, the result is just a different padding choice. There should be no difference in accuracy in padding left instead of right for example.
original code:
```python   
pad_total = [kernel_size_effective[i] - 1 for i in range(self.rank)]
```
code in the PR:
```python    
pad_total = [max(kernel_size_effective[i] - (self.strides[i]), 0) 
                 if (shape_in[i]%self.strides[i]) == 0 else
                 max(kernel_size_effective[i] - (shape_in[i]%self.strides[i]), 0) 
                 for i in range(self.rank)]
```
To have a really extensive test, you should compare the causal 3d convolution with standard TF implementation of same padding.

**Comment by Hyperparticle** [comment]
It looks like this change significantly impacts the accuracy of our pretrained models (MoViNet-A0-Stream 72% -> 68%). If this fix is applied, we may need to retrain all of the models to regain the accuracy. As such, it may be necessary to disable the fix by default until we can release the new weights.

**Comment by Atze00** [comment]
That's the expected behavior unfortunately, thank you for the update.

**Comment by Hyperparticle** [comment]
We have now trained new models that use the fixed padding. See https://github.com/tensorflow/models/commit/08f9393ab192a65a6be05793e2db5885963bc6d6.
